contrib/babelfishpg_tsql/src/iterative_exec.c

```diff
@@ -1132,8 +1135,12 @@ int dispatch_stmt_handle_error(PLtsql_execstate *estate,
 		 * We do not start savepoint for batch commands as
 		 * error handling must be taken care of at statement
 		 * level
+		 * For RO functions start savepoint even when transaction
+		 * is not active to retain top level portals. A transaction
+		 * rollback will cleanup portal data which can lead to
+		 * problems when control returns back to portal level
 		 */
-		if ((!pltsql_disable_internal_savepoint && !is_batch_command(stmt) && IsTransactionBlockActive()))
+		if (!pltsql_disable_internal_savepoint && !is_batch_command(stmt) && (IsTransactionBlockActive() || ro_func))
```

Comment:
Just for curiosity, does it mean we start N internal save points if a select statement contains N functions?

Reply:
if there are N function calls like N1, N2, N3 then it will become sp1 N1 relsp1 sp2 N2 relsp2 sp3 N3 relsp3. If function call are nested where N1 -> N2 -> N3, there will be 3 active savepoints